src/GoodsListElement/GoodsListElement.js

```diff
@@ -1,22 +1,128 @@
-        )
-    }
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      editing: false,
+      title: '',
+      weight: 0,
+      description: '',
+      category: '',
+    };
+
+    this.onEdit = (e) => {
+      e.stopPropagation();
+      const { item: { title, weight, description, category } } = this.props;
+      this.setState({
+        editing: true,
+        title,
+        weight,
+        description,
+        category,
+      });
+    };
+
+    this.onInputChange = ({ target }) => {
+      this.setState({
+        [target.name]: target.value,
+      });
+    };
+
+    this.onSave = () => {
+      const { title, weight, description, category } = this.state;
+      const { item: { id }, onSave } = this.props;
+
+      this.setState({
+        editing: false,
+      });
+      onSave(id, { title, weight, description, category });
+    };
+
+    this.onDelete = (e) => {
+      e.stopPropagation();
+      this.props.onDelete(this.props.item.id);
+    };
+
+    this.onToggle = () => {
+      if (!this.state.editing) {
+        this.props.onToggle(this.props.item.id);
+      }
+    };
+  }
+
+  render() {
+    const { item, selected, categories } = this.props;
+    const { title, weight, description, category = 'uncategorized' } = item;
+    const elementClassName = selected ?
+      'GoodsListElement isSelected' : 'GoodsListElement';
+    const titleColumnContent = this.state.editing ?
+      <input type="text"
+        defaultValue={ title }
+        name="title"
+        onChange={ this.onInputChange }
+      /> :
+      title;
+    const weightColumnContent = this.state.editing ?
+      <input type="number"
+        defaultValue={ weight }
+        name="weight"
+        onChange={ this.onInputChange }
+      /> :
+      weight;
+    const descriptionColumnContent = this.state.editing ?
+      <input type="text"
+        defaultValue={ description }
+        name="description"
+        onChange={ this.onInputChange }
+      /> :
+      description;
+    const categoryColumnContent = this.state.editing ?
+      <CategorySelect
+        categories={ categories }
+        onChange={ this.onInputChange }
+      /> :
+      getCategory(category, 'slug', categories).name;
+
+    return (
+      <div className={ elementClassName } onClick={ this.onToggle }>
+        <div className="GoodsListElement_Column">{ titleColumnContent }</div>
+        <div className="GoodsListElement_Column">{ weightColumnContent }</div>
+        <div className={ 'GoodsListElement_Column ' +
+          'GoodsListElement_ColumnDescription' }>
+          { descriptionColumnContent }
+        </div>
+        <div className="GoodsListElement_Column">{ categoryColumnContent }</div>
+        <div className="GoodsListElement_Column GoodsListElement_Button">
+          {
+            ( !this.state.editing &&
+              <button onClick={this.onEdit}>{ __('Edit') }</button>
+            ) || <button onClick={this.onSave}>{ __('Save') }</button>
```

Comment:
To read it easier I would write:
```
{ !this.state.editing && <button onClick={this.onEdit}>{ __('Edit') }</button>}
{ this.state.editing && <button onClick={this.onSave}>{ __('Save') }</button>}
```

Or you can use also:
`this.state.editing ? (<button onClick={this.onEdit}>{ __('Edit') }</button>) : (<button onClick={this.onSave}>{ __('Save') }</button>)`

But I prefer the first version or also good one:

```
onRowAction = () => {
  if(this.state.editing) {
    this.onSave()
  } else {
    this.onEdit()
  }
}
`<button onClick={this.onRowAction}>{ __(this.state.editing ? 'Save' : 'Edit' ) }</button>`
`...`
```

Reply:
Thanks, really neat solution! Appreciate that! :smile: